cmd/pitr/recoverer/recoverer.go

```diff
@@ -0,0 +1,252 @@
+	backupNAme       string
+	excludingGTIDSet string
+}
+
+type Config struct {
+	PXCServiceName string
+	PXCUser        string
+	PXCPass        string
+	S3Endpoint     string
+	S3AccessKeyID  string
+	S3AccessKey    string
+	S3BucketName   string
+	S3Region       string
+	RecoverTime    int64
+	RecoverType    string
+	BackupName     string
+}
+
+type RecoverType string
+
+func New(c Config) (*Recoverer, error) {
+	s3, err := storage.NewS3(c.S3Endpoint, c.S3AccessKeyID, c.S3AccessKey, c.S3BucketName, c.S3Region, true)
+	if err != nil {
+		return nil, errors.Wrap(err, "new storage manager")
+	}
+	return &Recoverer{
+		storage:        s3,
+		recoverTime:    c.RecoverTime,
+		pxcUser:        c.PXCUser,
+		pxcPass:        c.PXCPass,
+		pxcServiceName: c.PXCServiceName,
+		recoverType:    RecoverType(c.RecoverTime),
+		backupNAme:     c.BackupName,
+	}, nil
+}
+
+const (
+	Latest      RecoverType = "latest"
+	Date        RecoverType = "date"
+	Transaction RecoverType = "transaction"
+	Skip        RecoverType = "skip"
+)
+
+func (r *Recoverer) getHost() (string, error) {
+	cmd := exec.Command("peer-list", "-on-start=/usr/bin/get-pxc-state", "-service="+r.pxcServiceName)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", errors.Wrap(err, "get output")
+	}
+	nodes := strings.Split(string(out), "node:")
+	sort.Strings(nodes)
+	lastHost := ""
+	for _, node := range nodes {
+		if strings.Contains(node, "wsrep_ready:ON:wsrep_connected:ON:wsrep_local_state_comment:Synced:wsrep_cluster_status:Primary") {
+			nodeArr := strings.Split(node, ".")
+			lastHost = nodeArr[0]
+		}
+	}
+	if len(lastHost) == 0 {
+		return "", errors.New("cant find host")
+	}
+
+	return lastHost + "." + r.pxcServiceName, nil
+}
+
+func (r *Recoverer) Run() error {
+	startGTID, err := r.GetLastBackupGTID()
+	if err != nil {
+		return errors.Wrap(err, "get last gtid")
+	}
+	err = r.GetBinlogList(startGTID)
+	if err != nil {
+		return errors.Wrap(err, "get binlog list")
+	}
+	err = r.Recover()
+	if err != nil {
+		return errors.Wrap(err, "recover")
+	}
+	return nil
+}
+
+func (r *Recoverer) DownloadBinlogs() error {
+	for _, binlog := range r.storage.ListObjects("") {
+		if strings.Contains(binlog, "set") {
+			continue
+		}
+		data, err := r.storage.GetObject(binlog)
+		if err != nil {
+			return errors.Wrap(err, "get object")
+		}
+		err = ioutil.WriteFile(binlog, data, 0644)
+		if err != nil {
+			return errors.Wrap(err, "write file")
+		}
+		r.binlogs = append(r.binlogs, binlog)
+	}
+	return nil
+}
+
+func (r *Recoverer) Recover() error {
+	host, err := r.getHost()
+	if err != nil {
+		return errors.Wrap(err, "get host")
+	}
+
+	for _, binlog := range r.binlogs {
+		err = r.DownloadBinlog(binlog)
+		if err != nil {
+			return errors.Wrap(err, "download binlog")
+		}
+		flags := ""
+
+		// TODO: add logic for all types
+		switch r.recoverType {
+		case Skip:
+			flags = "--exclude-gtids=" + r.excludingGTIDSet
+		case Transaction:
+		case Date:
+		case Latest:
+		}
```

Comment:
make no sense to calculate it for each file - it remains the same

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/673/commits/74347e94f05b3b59a0399735a7a9e3ea4ae3d768